py4web/core.py

```diff
@@ -872,44 +864,46 @@ def decorator(func):
 
             if Fixture.__fixture_debug__:
                 # in debug mode log all calls to fixtures
-                def call(obj, f, args=()):
-                    logging.debug("Calling %s.%s", obj.__class__.__name__, f)
-                    return getattr(obj, f)(*args)
+                def call(f, context):
+                    logging.debug(
+                        f"Calling {f.__self__.__class__.__name__}.{f.__name__}"
+                    )
+                    return f(context)
 
             else:
 
-                def call(obj, f, args=()):
-                    return getattr(obj, f)(*args)
+                def call(f, context):
+                    return f(context)
 
             @functools.wraps(func)
             def wrapper(*args, **kwargs):
                 # data shared by all fixtures in the pipeline for each request
-                shared_data = {"template_context": {}}
+                processed = []
+                context = {
+                    "fixtures": fixtures,
+                    "status": 200,
+                    "output": None,
+                    "exception": None,
+                    "processed": processed,
+                }
                 try:
-                    for obj in fixtures:
-                        call(obj, "on_request")
-                    ret = func(*args, **kwargs)
-                    for obj in fixtures:
-                        ret = call(obj, "transform", (ret, shared_data))
-                    for obj in fixtures:
-                        call(obj, "on_success", (200,))
-                    return ret
+                    for fixture in fixtures:
+                        call(fixture.on_request, context)
+                        processed.append(fixture)
+                    context["output"] = func(*args, **kwargs)
                 except HTTP as http:
-                    if http.type == http.Type.success:
-                        for obj in fixtures:
-                            call(obj, "on_success", (http.status,))
+                    context["status"] = http.status
+                    raise http
```

Comment:
What about fixtures postprocessing?

Reply:
we discussed this on the google group. raise HTTP(status, output) is equivalent to response.status=status and return output. So all fixture.on_success are expected to be called except they will see output=="" because the actually output is stored into the HTTP and cannot be modified by the fixtures.